File: 백트래킹/스도쿠/baekjoon-2580.py
def square(y, x):
    return (y // 3) * 3 + (x // 3)


def go(index):
    if index == 81:
        for row in a:
            print(' '.join(map(str, row)))
        return

    y = index // n
    x = index % n

    if a[y][x] != 0:
        go(index + 1)
    else:
        for i in range(1, 10):
            if check_col[x][i] is False and check_row[y][i] is False and \
                    check_square[square(y, x)][i] is False:
                check_col[x][i] = check_row[y][i] = check_square[square(y, x)][
                    i] = True
                a[y][x] = i

                go(index + 1)

                check_col[x][i] = check_row[y][i] = check_square[square(y, x)][
                    i] = False
                a[y][x] = 0


n = 9
a = [list(map(int, input().split())) for _ in range(n)]

check_col = [[False] * 10 for _ in range(n)]
check_row = [[False] * 10 for _ in range(n)]
check_square = [[False] * 10 for _ in range(n)]

for row in range(n):
    for col in range(n):
        if a[row][col] != 0:
            check_col[col][a[row][col]] = True
            check_row[row][a[row][col]] = True
            check_square[square(row, col)][a[row][col]] = True

go(0)

File: 백트래킹/스도쿠/baekjoon-2580.cpp
#include <iostream>
#include <vector>
#include <utility>

using namespace std;

int a[10][10];
bool c[10][10];
bool c2[10][10];
bool c3[10][10];

vector<pair<int, int>> zero_pos;

int square(int row, int col) {
  return (row / 3) * 3 + (col / 3);
}

void go(int index) {
  if(index == (int)zero_pos.size()) {
    for(int i=0; i<9; i++) {
      for(int j=0; j<9; j++) {
        cout << a[i][j] << " ";
      }
      cout << endl;
    }
    cout << endl;
    exit(0);
  }

  int y = zero_pos[index].first;
  int x = zero_pos[index].second;
  for(int number=1; number <= 9; number++) {
    if(c[y][number] == 0 && c2[x][number] == 0 && c3[square(y, x)][number] == 0) {
      c[y][number] = c2[x][number] = c3[square(y, x)][number] = true;
      a[y][x] = number;
      go(index + 1);
      c[y][number] = c2[x][number] = c3[square(y, x)][number] = false;
      a[y][x] = 0;
    }
  }
}

int main()
{
  for(int i=0; i<9; i++) {
    for(int j=0; j<9; j++) {
      cin >> a[i][j];
      if(a[i][j] != 0) {
        c[i][a[i][j]] = true;
        c2[j][a[i][j]] = true;
        c3[square(i, j)][a[i][j]] = true;
      }
      else {
        zero_pos.push_back(make_pair(i, j));
      }
    }
  }

  go(0);
  return 0;
}